docs/docs/remark-ast.md

```diff
@@ -0,0 +1,202 @@
+
+  return markdownAST
+}
+```
+
+The first parameter is all of the default properties that can be used in plugins (actions, store, getNodes, schema, etc.) plus a couple just for gatsby-transformer-remark plugins. The most relevant field for our purposes is the `markdownAST` field which is destructured in the code snippet above.
+
+As with other Gatsby plugins, the 2nd parameter is the `pluginOptions` which is obtained from the definition in `gatsby-config.js` file.
+
+Finally, the function will return the `markdownAST` after the fields you wish to be edited are transformed.
+
+## Adding the plugin to your site
+
+You likely will want to grab `gatsby-source-filesystem` to inject the file nodes into Gatsby's GraphQL schema. In this example it is assumed that the Markdown files exist in a `src/data/` directory.
+
+The plugin is now initially set so you can add it as a sub-plugin inside `gatsby-transformer-remark`
+
+```js:title=gatsby-config.js
+module.exports = {
+  plugins: [
+    {
+      resolve: "gatsby-source-filesystem",
+      options: {
+        name: `data`,
+        path: `${__dirname}/src/data/`,
+      },
+    },
+    {
+      resolve: `gatsby-transformer-remark`,
+      options: {
+        plugins: [`gatsby-remark-purple-headers`],
+      },
+    },
+  ],
+}
+```
+
+If you want to add some options, you could switch to the object syntax:
+
+```js
+{
+  resolve: `gatsby-remark-purple-headers`,
+  options: {
+    // Options here
+  }
+}
+```
+
+## Find and Modify Markdown Nodes
+
+When modifying nodes, you'll want to walk the tree and then implement new functionality on specific nodes.
+
+A node module to help with is [unist-util-visit](https://github.com/syntax-tree/unist-util-visit), a walker for `unist` nodes. For reference, Unist (Unified Syntax Tree) is a standard for Markdown syntax trees and parsers that include well known parsers in the Gatsby world like Remark and MDX.
+
+As an example from `unist-util-visit`'s readme, it allows for an easy interface to visit particular nodes based on a particular type:
+
+```js
+var remark = require("remark")
+var visit = require("unist-util-visit")
+
+var tree = remark().parse("Some _emphasis_, **importance**, and `code`.")
+
+visit(tree, "text", visitor)
+
+function visitor(node) {
+  console.log(node)
+}
+```
+
+Here, it finds all text nodes and will `console.log` the nodes.
+
+With this technique in mind, you can similarly traverse the AST from your plugin and add additional functionality, like so:
+
+```js:title=plugins/gatsby-remark-purple-headers/index.js
+const visit = require("unist-util-visit")
+
+module.exports = ({ markdownAST }, pluginOptions) => {
+  // highlight-next-line
+  visit(markdownAST, "heading", node => {
+    // Do stuff with heading nodes
+  })
+
+  return markdownAST
+}
+```
+
+Next, by visiting all heading nodes and passing them into a transformer function, you can manipulate the particular nodes to match your usecase.
+
+Looking again at the AST node for header:
```

Comment:
```suggestion
Looking again at the AST node for heading:
```
Looking at the JSON structure and nature of the tutorial, "heading" is probably more accurate than "header" in an HTML / markup context.